// @neo4j/graphql/dist/translate/cypher-builder/statements/Create.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Create = void 0;
const MatchPattern_1 = require("../MatchPattern");
const Param_1 = require("../references/Param");
const Query_1 = require("./Query");
const Return_1 = require("./Return");
class Create extends Query_1.Query {
    constructor(node, params = {}, parent) {
        super(parent);
        this.node = node;
        this.matchPattern = new MatchPattern_1.MatchPattern(node).withParams(params);
        this.setParams = {};
    }
    set(params) {
        this.setParams = params;
        return this;
    }
    cypher(context, childrenCypher) {
        const nodeCypher = this.matchPattern.getCypher(context);
        return `CREATE ${nodeCypher}\n${this.composeSet(context)}\n${childrenCypher}`;
    }
    return(...args) {
        const returnStatement = new Return_1.ReturnStatement(this, args);
        this.addStatement(returnStatement);
        return this;
    }
    composeSet(context) {
        const nodeAlias = context.getVariableId(this.node);
        const setParams = Object.entries(this.setParams).map(([key, value]) => {
            return `${nodeAlias}.${key} = ${value instanceof Param_1.Param ? value.getCypher(context) : value}`;
        });
        if (setParams.length === 0)
            return "";
        return `SET ${setParams.join(",\n")}`;
    }
}
exports.Create = Create;
