// @neo4j/graphql/dist/translate/translate-read.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const neo4j_driver_1 = require("neo4j-driver");
const graphql_relay_1 = require("graphql-relay");
const create_projection_and_params_1 = __importDefault(require("./create-projection-and-params"));
const create_auth_and_params_1 = __importDefault(require("./create-auth-and-params"));
const constants_1 = require("../constants");
const create_connection_and_params_1 = __importDefault(require("./connection/create-connection-and-params"));
const create_interface_projection_and_params_1 = __importDefault(require("./create-interface-projection-and-params"));
const translate_top_level_match_1 = __importDefault(require("./translate-top-level-match"));
function translateRead({ node, context, isRootConnectionField, }) {
    const { resolveTree } = context;
    const varName = "this";
    let matchAndWhereStr = "";
    let authStr = "";
    let projAuth = "";
    let projStr = "";
    let cypherSort = false;
    const afterInput = resolveTree.args.after;
    const firstInput = resolveTree.args.first;
    const sortInput = resolveTree.args.sort;
    const optionsInput = (resolveTree.args.options || {});
    let limitStr = "";
    let offsetStr = "";
    let sortStr = "";
    let cypherParams = {};
    const connectionStrs = [];
    const interfaceStrs = [];
    const returnStrs = [];
    const hasLimit = Boolean(optionsInput?.limit) || optionsInput?.limit === 0;
    if (node.queryOptions) {
        optionsInput.limit = node.queryOptions.getLimit(optionsInput.limit);
    }
    const topLevelMatch = (0, translate_top_level_match_1.default)({ node, context, varName, operation: "READ" });
    matchAndWhereStr = topLevelMatch[0];
    cypherParams = { ...cypherParams, ...topLevelMatch[1] };
    const projection = (0, create_projection_and_params_1.default)({
        node,
        context,
        resolveTree,
        varName,
        isRootConnectionField,
    });
    [projStr] = projection;
    cypherParams = { ...cypherParams, ...projection[1] };
    if (projection[2]?.authValidateStrs?.length) {
        projAuth = `CALL apoc.util.validate(NOT (${projection[2].authValidateStrs.join(" AND ")}), "${constants_1.AUTH_FORBIDDEN_ERROR}", [0])`;
    }
    if (projection[2]?.connectionFields?.length) {
        projection[2].connectionFields.forEach((connectionResolveTree) => {
            const connectionField = node.connectionFields.find((x) => x.fieldName === connectionResolveTree.name);
            const connection = (0, create_connection_and_params_1.default)({
                resolveTree: connectionResolveTree,
                field: connectionField,
                context,
                nodeVariable: varName,
            });
            connectionStrs.push(connection[0]);
            cypherParams = { ...cypherParams, ...connection[1] };
        });
    }
    if (projection[2]?.interfaceFields?.length) {
        const prevRelationshipFields = [];
        projection[2].interfaceFields.forEach((interfaceResolveTree) => {
            const relationshipField = node.relationFields.find((x) => x.fieldName === interfaceResolveTree.name);
            const interfaceProjection = (0, create_interface_projection_and_params_1.default)({
                resolveTree: interfaceResolveTree,
                field: relationshipField,
                context,
                nodeVariable: varName,
                withVars: prevRelationshipFields,
            });
            prevRelationshipFields.push(relationshipField.dbPropertyName || relationshipField.fieldName);
            interfaceStrs.push(interfaceProjection.cypher);
            cypherParams = { ...cypherParams, ...interfaceProjection.params };
        });
    }
    const allowAndParams = (0, create_auth_and_params_1.default)({
        operations: "READ",
        entity: node,
        context,
        allow: {
            parentNode: node,
            varName,
        },
    });
    if (allowAndParams[0]) {
        cypherParams = { ...cypherParams, ...allowAndParams[1] };
        authStr = `CALL apoc.util.validate(NOT (${allowAndParams[0]}), "${constants_1.AUTH_FORBIDDEN_ERROR}", [0])`;
    }
    let cypher = [];
    if (isRootConnectionField) {
        const hasAfter = Boolean(afterInput);
        const hasFirst = Boolean(firstInput);
        const hasSort = Boolean(sortInput && sortInput.length);
        const sortCypherFields = projection[2]?.rootConnectionCypherSortFields ?? [];
        const sortCypherProj = sortCypherFields.map(({ alias, apocStr }) => `${alias}: ${apocStr}`);
        if (hasAfter && typeof afterInput === "string") {
            const offset = (0, graphql_relay_1.cursorToOffset)(afterInput) + 1;
            if (offset && offset !== 0) {
                offsetStr = `SKIP $${varName}_offset`;
                cypherParams[`${varName}_offset`] = (0, neo4j_driver_1.int)(offset);
            }
        }
        if (hasFirst) {
            limitStr = `LIMIT $${varName}_limit`;
            cypherParams[`${varName}_limit`] = firstInput;
        }
        if (hasSort) {
            const sortArr = sortInput.reduce((res, sort) => {
                return [
                    ...res,
                    ...Object.entries(sort).map(([field, direction]) => {
                        // if the sort arg is a cypher field, substitaute "edges" for varName
                        const varOrEdgeName = sortCypherFields.find((x) => x.alias === field) ? "edges" : varName;
                        return `${varOrEdgeName}.${field} ${direction}`;
                    }),
                ];
            }, []);
            sortStr = `ORDER BY ${sortArr.join(", ")}`;
        }
        returnStrs.push(`WITH COLLECT({ node: ${varName} ${projStr} }) as edges, totalCount`);
        returnStrs.push(`RETURN { edges: edges, totalCount: totalCount } as ${varName}`);
        cypher = [
            "CALL {",
            matchAndWhereStr,
            authStr,
            ...(projAuth ? [`WITH ${varName}`, projAuth] : []),
            `WITH COLLECT(this) as edges`,
            `WITH edges, size(edges) as totalCount`,
            `UNWIND edges as ${varName}`,
            `WITH ${varName}, totalCount, { ${sortCypherProj.join(", ")}} as edges`,
            `RETURN ${varName}, totalCount, edges`,
            ...(sortStr ? [sortStr] : []),
            ...(offsetStr ? [offsetStr] : []),
            ...(limitStr ? [limitStr] : []),
            "}",
            ...connectionStrs,
            ...interfaceStrs,
            ...returnStrs,
        ];
    }
    else {
        const hasOffset = Boolean(optionsInput.offset) || optionsInput.offset === 0;
        if (hasOffset) {
            offsetStr = `SKIP $${varName}_offset`;
            cypherParams[`${varName}_offset`] = optionsInput.offset;
        }
        if (optionsInput.limit) {
            limitStr = `LIMIT $${varName}_limit`;
            cypherParams[`${varName}_limit`] = optionsInput.limit;
        }
        if (optionsInput.sort && optionsInput.sort.length) {
            const sortArr = optionsInput.sort.reduce((res, sort) => {
                return [
                    ...res,
                    ...Object.entries(sort).map(([field, direction]) => {
                        if (!cypherSort && node.cypherFields.some((f) => f.fieldName === field)) {
                            cypherSort = true;
                        }
                        return `${varName}.${field} ${direction}`;
                    }),
                ];
            }, []);
            sortStr = `ORDER BY ${sortArr.join(", ")}`;
        }
        returnStrs.push(`RETURN ${varName} ${projStr} as ${varName}`);
        const projectCypherFieldsAfterLimit = node.cypherFields.length && hasLimit && !cypherSort;
        if (projectCypherFieldsAfterLimit) {
            cypher = [
                "CALL {",
                matchAndWhereStr,
                authStr,
                ...(projAuth ? [`WITH ${varName}`, projAuth] : []),
                `RETURN ${varName}`,
                ...(sortStr ? [sortStr] : []),
                ...(offsetStr ? [offsetStr] : []),
                ...(limitStr ? [limitStr] : []),
                "}",
                ...connectionStrs,
                ...interfaceStrs,
                ...returnStrs,
            ];
        }
        else {
            cypher = [
                matchAndWhereStr,
                authStr,
                ...(projAuth ? [`WITH ${varName}`, projAuth] : []),
                ...connectionStrs,
                ...interfaceStrs,
                ...returnStrs,
                ...(sortStr ? [sortStr] : []),
                ...(offsetStr ? [offsetStr] : []),
                ...(limitStr ? [limitStr] : []),
            ];
        }
    }
    return [cypher.filter(Boolean).join("\n"), cypherParams];
}
exports.default = translateRead;
